File: libs/js/src/main/scripts/ambra/savedSearch.js
/*
 * $HeadURL$
 * $Id$
 * Copyright (c) 2006-2012 by Public Library of Science http://plos.org http://ambraproject.org
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
dojo.provide("ambra.savedSearch");
dojo.require("ambra.general");
dojo.require("ambra.domUtil");
dojo.require("ambra.formUtil");

ambra.savedSearch = {

  init: function() {
  },

  /**
   * clicking on save search link will show the dialog box.
   */
  show: function(action){
    console.log("saved search show");
    var searchEdit = dojo.byId("searchEdit");
    if (searchEdit && searchEdit.value) {
      dojo.byId("text_name_savedsearch").value = searchEdit.value;
    }

    dojo.byId("span_error_savedsearch").innerHTML = "";

    _savedSearchDlg.show();
    return false;
  },

  /**
   * clicking on cancel button will hide the dialog box.
   */
  hide: function() {
    console.log("saved search hide");
    dojo.byId("span_error_savedsearch").innerHTML = "";
    _savedSearchDlg.hide();
  },

  /**
   * clicking on save button will save the form data.
   */
  save: function(success_callback) {
    //Copy values up to the parent form
    dojo.byId("searchName").value = dojo.byId("text_name_savedsearch").value;
    dojo.byId("weekly").value = dojo.byId("cb_weekly_savedsearch").checked;
    dojo.byId("monthly").value = dojo.byId("cb_monthly_savedsearch").checked;

    var xhrArgs = {
      url: "/search/saveSearch.action",
      //Grab all for form values from the parent form
      form: dojo.byId("searchFormOnSearchResultsPage"),
      //postData: JSON.stringify(content),
      handleAs:'json-comment-filtered',
      load: function(response, ioArgs) {
        if(response.exception) {
          var errorMessage = response.exception.message;

          dojo.byId("span_error_savedsearch").innerHTML = "Exception: " + errorMessage;

          return;
        }

        if (response.actionErrors && response.actionErrors.length > 0) {
          //The action in question can only return one message
          var errorMessage = response.actionErrors[0];

          dojo.byId("span_error_savedsearch").innerHTML = "Error: " + errorMessage;
          return;
        }

        _savedSearchDlg.hide();
      },
      error: function(response, ioArgs) {
        dojo.byId("span_error_savedsearch").innerHTML = response;
        console.log("failed with error: " + response);
      }
    }

    dojo.xhrPost(xhrArgs);
  }
}

File: models/src/main/java/org/ambraproject/models/SavedSearch.java
/*
 * $HeadURL$
 * $Id$
 * Copyright (c) 2006-2012 by Public Library of Science http://plos.org http://ambraproject.org
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.ambraproject.models;

import java.util.Calendar;
import java.util.Date;

/**
 * POJO Object to contain all the information about saved searches.
 * It is mapped with table savedSearch.
 */
public class SavedSearch extends AmbraEntity {

  private String searchName;
  private SavedSearchQuery searchQuery;
  private Date lastWeeklySearchTime;
  private Date lastMonthlySearchTime;
  private boolean weekly;
  private boolean monthly;

  //TODO: Define type, "User Defined" or "Journal Alert"

  public SavedSearch() {
    super();
    lastWeeklySearchTime = Calendar.getInstance().getTime();
    lastMonthlySearchTime= Calendar.getInstance().getTime();
  }

  public SavedSearch(String searchName, SavedSearchQuery searchQuery) {
    this();
    this.searchName = searchName;
    this.searchQuery = searchQuery;
  }
  
  /**
   * getter for search name
   * @return searchName
   */
  public String getSearchName() {
    return searchName;
  }

  /**
   * setter for search name
   * @param searchName
   */
  public void setSearchName(String searchName) {
    this.searchName = searchName;
  }

  public SavedSearchQuery getSearchQuery() {
    return searchQuery;
  }

  public void setSearchQuery(SavedSearchQuery searchQuery) {
    this.searchQuery = searchQuery;
  }

  public Date getLastWeeklySearchTime() {
    return lastWeeklySearchTime;
  }

  public void setLastWeeklySearchTime(Date lastWeeklySearchTime) {
    this.lastWeeklySearchTime = lastWeeklySearchTime;
  }

  public Date getLastMonthlySearchTime() {
    return lastMonthlySearchTime;
  }

  public void setLastMonthlySearchTime(Date lastMonthlySearchTime) {
    this.lastMonthlySearchTime = lastMonthlySearchTime;
  }

  /**
   * @return weekly
   */
  public boolean getWeekly() {
    return weekly;
  }

  /**
   * setter for weekly
   * @param weekly
   */
  public void setWeekly(boolean weekly) {
    this.weekly = weekly;
  }

  /**
   * @return monthly
   */
  public boolean getMonthly() {
    return monthly;
  }

  /**
   * setter for monthly
   * @param monthly
   */
  public void setMonthly(boolean monthly) {
    this.monthly = monthly;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (!(o instanceof SavedSearch)) return false;

    SavedSearch that = (SavedSearch) o;

    if (getID() != null ? !getID().equals(that.getID()) : that.getID() != null) return false;
    if (searchName != null ? !searchName.equals(that.searchName) : that.searchName != null) return false;
    if (searchQuery != null ? !searchQuery.equals(that.searchQuery) : that.searchQuery != null) return false;
    return true;
  }

  @Override
  public int hashCode() {
    int result = getID() != null ? getID().hashCode() : 0;
    result = 31 * result + (searchName != null ? searchName.hashCode() : 0);
    result = 31 * result + (searchQuery != null ? searchQuery.hashCode() : 0);
    return result;
  }
}
